arbnode/resource_management.go

```diff
@@ -0,0 +1,171 @@
+	MemoryLimitPercent: 0,
+}
+
+func ResourceManagementConfigAddOptions(prefix string, f *flag.FlagSet) {
+	f.Int(prefix+".mem-limit-percent", DefaultResourceManagementConfig.MemoryLimitPercent, "RPC calls are throttled if system memory utilization exceeds this percent value, zero (default) is disabled")
+}
+
+type resourceManagementHttpServer struct {
+	inner http.Handler
+	c     limitChecker
+}
+
+func newResourceManagementHttpServer(inner http.Handler, c limitChecker) *resourceManagementHttpServer {
+	return &resourceManagementHttpServer{inner: inner, c: c}
+}
+
+func (s *resourceManagementHttpServer) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	exceeded, err := s.c.isLimitExceeded()
+	if err != nil {
+		log.Error("Error checking memory limit", "err", err, "checker", s.c)
+	} else if exceeded {
+		http.Error(w, "Too many requests", http.StatusTooManyRequests)
+		return
+	}
+
+	log.Info("Limit not exceeded, serving request.")
+	s.inner.ServeHTTP(w, req)
+}
+
+type limitChecker interface {
+	isLimitExceeded() (bool, error)
+	String() string
+}
+
+func newLimitChecker(conf *ResourceManagementConfig) limitChecker {
+	{
+		c := newCgroupsV1MemoryLimitChecker(DefaultCgroupsV1MemoryDirectory, conf.MemoryLimitPercent)
+		if isSupported(c) {
+			log.Info("Cgroups v1 detected, enabling memory limit RPC throttling")
+			return c
+		}
+	}
+
+	log.Error("No method for determining memory usage and limits was discovered, disabled memory limit RPC throttling")
+	return &trivialLimitChecker{}
+}
+
+type trivialLimitChecker struct{}
+
+func (_ trivialLimitChecker) isLimitExceeded() (bool, error) {
+	return false, nil
+}
+
+func (_ trivialLimitChecker) String() string { return "trivial" }
+
+const DefaultCgroupsV1MemoryDirectory = "/sys/fs/cgroup/memory/"
+
+type cgroupsV1MemoryLimitChecker struct {
+	cgroupDir          string
+	memoryLimitPercent int
+
+	limitFile, usageFile, statsFile string
+}
+
+func newCgroupsV1MemoryLimitChecker(cgroupDir string, memoryLimitPercent int) *cgroupsV1MemoryLimitChecker {
+	return &cgroupsV1MemoryLimitChecker{
+		cgroupDir:          cgroupDir,
+		memoryLimitPercent: memoryLimitPercent,
+		limitFile:          cgroupDir + "/memory.limit_in_bytes",
+		usageFile:          cgroupDir + "/memory.usage_in_bytes",
+		statsFile:          cgroupDir + "/memory.stat",
+	}
+}
+
+func isSupported(c limitChecker) bool {
+	_, err := c.isLimitExceeded()
+	return err == nil
+}
+
+func (c *cgroupsV1MemoryLimitChecker) isLimitExceeded() (bool, error) {
+	var limit, usage, inactive int
+	var err error
+	limit, err = c.getIntFromFile(c.limitFile)
+	if err != nil {
+		return false, err
+	}
+	usage, err = c.getIntFromFile(c.usageFile)
+	if err != nil {
+		return false, err
+	}
+	inactive, err = c.getInactive()
+	if err != nil {
+		return false, err
+	}
+	return usage-inactive >= ((limit * c.memoryLimitPercent) / 100), nil
+}
+
+func (c cgroupsV1MemoryLimitChecker) getIntFromFile(fileName string) (int, error) {
```

Comment:
nit: drop get prefix, same with getInactive.

https://google.github.io/styleguide/go/decisions#getters

Reply:
Should it have a prefix like `read` instead since it is reading from a file? I understand that accessors shouldn't have a prefix, but is this an accessor?